src/anyio/functools.py

```diff
@@ -0,0 +1,253 @@
+from .lowlevel import RunVar
+
+if sys.version_info >= (3, 11):
+    from typing import ParamSpec
+else:
+    from typing_extensions import ParamSpec
+
+T = TypeVar("T")
+S = TypeVar("S")
+P = ParamSpec("P")
+lru_cache_items: RunVar[
+    defaultdict[Callable[..., Any], tuple[OrderedDict[Hashable, Any], Lock]]
+] = RunVar("lru_cache_items")
+
+
+class _InitialMissingType:
+    pass
+
+
+initial_missing: _InitialMissingType = _InitialMissingType()
+
+
+class AsyncCacheInfo(NamedTuple):
+    hits: int
+    misses: int
+    maxsize: int | None
+    currsize: int
+
+
+class AsyncCacheParameters(TypedDict):
+    maxsize: int | None
+    typed: bool
+
+
+class AsyncLRUCacheWrapper(Generic[P, T]):
+    def __init__(
+        self, func: Callable[..., Awaitable[T]], maxsize: int | None, typed: bool
+    ):
+        self.__wrapped__ = func
+        self._hits: int = 0
+        self._misses: int = 0
+        self._maxsize = max(maxsize, 0) if maxsize is not None else None
+        self._currsize: int = 0
+        self._typed = typed
+        update_wrapper(self, func)
+
+    def cache_info(self) -> AsyncCacheInfo:
+        return AsyncCacheInfo(self._hits, self._misses, self._maxsize, self._currsize)
+
+    def cache_parameters(self) -> AsyncCacheParameters:
+        return AsyncCacheParameters(maxsize=self._maxsize, typed=self._typed)
+
+    def cache_clear(self) -> None:
+        if cache := lru_cache_items.get(None):
+            cache.pop(self.__wrapped__, None)
+            self._hits = self._misses = self._currsize = 0
+
+    async def __call__(self, *args: P.args, **kwargs: P.kwargs) -> T:
+        # Easy case first: if maxsize == 0, no caching is done
+        if self._maxsize == 0:
+            value = await self.__wrapped__(*args, **kwargs)
+            self._misses += 1
+            return value
+
+        # The key is constructed as a flat tuple to avoid memory overhead
+        key: tuple[Any, ...] = args
+        if kwargs:
+            # initial_missing is used as a separator
+            key += (initial_missing,) + sum(kwargs.items(), ())
+
+        if self._typed:
+            key += tuple(type(arg) for arg in args)
+            if kwargs:
+                key += (initial_missing,) + tuple(type(val) for val in kwargs.values())
+
+        try:
+            cache = lru_cache_items.get()
+        except LookupError:
+            cache = defaultdict(lambda: (OrderedDict(), Lock()))
+            lru_cache_items.set(cache)
+
+        cache_entry, lock = cache[self.__wrapped__]
+        async with lock:
+            try:
+                value = cache_entry[key]
+            except KeyError:
+                self._misses += 1
+                if self._maxsize is not None and len(cache_entry) >= self._maxsize:
+                    cache_entry.popitem(last=False)
+                else:
+                    self._currsize += 1
+
+                cache_entry[key] = value = await self.__wrapped__(*args, **kwargs)
```

Comment:
I don't think that `self.__wrapped__`'s cache's lock should be held while `__wrapped__` is called. This causes all callers of the function to go single-file. IIUC stdlib `lru_cache` does not hold its lock while calling the user function.